src/memory/AmdFlash.cc

```diff
@@ -35,152 +41,125 @@ AmdFlash::AmdFlash(const std::string& name, const ValidatedChip& validatedChip,
-				rom->getFilename());
-		} catch (MSXException& e) {
-			// ignore error
-			assert(rom == nullptr); // 'rom' remains nullptr
-			// only if an actual sha1sum was given, tell the user we failed to use it
-			if (initialContentSpecified) {
-				config.getCliComm().printWarning(
-					"Could not load specified initial content "
-					"for flash ROM: ", e.getMessage());
-			}
+			file = File(path, File::OpenMode::LOAD_PERSISTENT);
+		} catch (FileException& e) {
+			config.getCliComm().printWarning("Couldn't load Flash data ",
+				filename, " (", e.getMessage(), ").");
 		}
+	} catch (FileException& e) {
+		// SRAM file not found, assuming blank SRAM content.
 	}
 
-	readAddress.resize(numSectors);
-	auto romSize = rom ? rom->size() : 0;
-	size_t offset = 0;
-	for (size_t sector = 0; const AmdFlash::Region& region : chip.geometry.regions) {
-		for (size_t regionSector = 0; regionSector < region.count; regionSector++, sector++) {
-			auto sectorSize = region.size;
-			if (isSectorWritable(sector)) {
-				readAddress[sector] = &(*ram)[writeAddress[sector]];
-				if (!loaded) {
-					auto* ramPtr = const_cast<uint8_t*>(
-						&(*ram)[writeAddress[sector]]);
-					if (offset >= romSize) {
-						// completely past end of rom
-						ranges::fill(std::span{ramPtr, sectorSize}, 0xFF);
-					} else if (offset + sectorSize >= romSize) {
-						// partial overlap
-						auto last = romSize - offset;
-						auto missing = sectorSize - last;
-						const uint8_t* romPtr = &(*rom)[offset];
-						ranges::copy(std::span{romPtr, last}, ramPtr);
-						ranges::fill(std::span{&ramPtr[last], missing}, 0xFF);
-					} else {
-						// completely before end of rom
-						const uint8_t* romPtr = &(*rom)[offset];
-						ranges::copy(std::span{romPtr, sectorSize}, ramPtr);
-					}
-				}
+	size_t fileRemaining = file.is_open() ? file.getSize() : 0;
+	const std::span<const uint8_t> romData = rom ? std::span{*rom} : std::span<const uint8_t>{};
+	const std::span<uint8_t> ramData = ram.getWriteBackdoor();
+	for (size_t sector = 0, address = 0; const AmdFlash::Region& region : chip.geometry.regions) {
+		for (size_t regionSector = 0; regionSector < region.count; regionSector++, sector++, address += region.size) {
+			if (sectors[sector].modified) {
+				const size_t size = std::min<size_t>(region.size, fileRemaining);
+				if (size > 0) file.read(ramData.subspan(address, size));
+				ranges::fill(ramData.subspan(address + size, region.size - size), 0xFF);
+				fileRemaining -= size;
 			} else {
-				assert(rom); // must have rom constructor parameter
-				if ((offset + sectorSize) <= romSize) {
-					readAddress[sector] = &(*rom)[offset];
-				} else {
-					readAddress[sector] = nullptr;
-				}
+				const size_t size = std::min(address + region.size, romData.size()) - std::min(address, romData.size());
+				ranges::copy(romData.subspan(address, size), ramData.subspan(address, size));
+				ranges::fill(ramData.subspan(address + size, region.size - size), 0xFF);
 			}
```

Comment:
In each iteration you fill an extra part of `ramData`, and the `then` and the `else` branch fill the same part. This is one of the things I verified during my review. So it might be worth making this explicit, by moving the sub-expression `ramData.subspan(address, size)` before the `if`. Also both branches fill the remainder of this part with 0xff, could that be moved after the `if`?